### PR TITLE
feat(TextArea): Allow optional character counting and limit

### PR DESCRIPTION
**Pull request by katiezutter**
### Overview

TextArea optionally displays the current character count and the character limit.

The [figma designs](https://www.figma.com/file/PG0kVYDIuLk3hE566beZUV/Profiles-(Concept-Refinement)?node-id=1226%3A0) originally called for the count to be inside the textarea; I figured that might be a little tricky to implement especially since the box area is expandable, but if anyone has a quick answer for this I'm curious! Otherwise, Timea said this is okay as it is.

Screen Shot 2021-05-28 at 2 14 17 PM

To see the changes in the real world, check out the preview environment for this monolith PR: https://github.com/codecademy-engineering/Codecademy/pull/24842

`https://pr-24842.dev-eks.codecademy.com/profiles/<your-username>/edit`

### PR Checklist

- [x] Related to designs: [Profiles figma](https://www.figma.com/file/PG0kVYDIuLk3hE566beZUV/Profiles-(Concept-Refinement)?node-id=1226%3A0)
- [x] Related to JIRA ticket: [EGG-1271](https://codecademy.atlassian.net/browse/EGG-1271)
- [x] I have run this code to verify it works
- [x] This PR includes unit tests for the code change

### CODE REVIEW

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://60b01d292b7d94484e408e0c--gamut-preview.netlify.app>https://60b01d292b7d94484e408e0c--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/60b01d292b7d94484e408e0c>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.4.1-alpha.cc89b4.0<br/>@codecademy/gamut-labs@18.3.1-alpha.cc89b4.0<br/>@codecademy/gamut@33.2.1-alpha.cc89b4.0<br/>@codecademy/styleguide@40.1.1-alpha.cc89b4.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://60b131dec053c86a9fdd3776--gamut-preview.netlify.app>https://60b131dec053c86a9fdd3776--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/60b131dec053c86a9fdd3776>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.4.3-alpha.38f379.0<br/>@codecademy/gamut-labs@18.3.3-alpha.38f379.0<br/>@codecademy/gamut@33.4.1-alpha.38f379.0<br/>@codecademy/styleguide@40.3.1-alpha.38f379.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://60b138908186080b8a43d76e--gamut-preview.netlify.app>https://60b138908186080b8a43d76e--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/60b138908186080b8a43d76e>Deploy Logs</a></p>

File: packages/gamut/src/Form/TextArea.tsx [hunk 1]
```diff
@@ -28,24 +31,48 @@ const StyledTextArea = styled.textarea<TextAreaProps>`
...
-        error={error}
-        activated={activated}
-        onChange={(event) => changeHandler(event)}
-      />
+      <>
+        <StyledTextArea
+          {...rest}
+          aria-label={
+            characterMax ? `maximum of ${characterMax} characters` : ''
+          }
```
Comment (katiezutter) [inline packages/gamut/src/Form/TextArea.tsx:59]: I really wasn't sure what to do in terms of accessibility: does anyone know what the best practice is here for a character count? I was seeing some advice to use `aria-live`, and then some advice that it's too annoying to hear the character count each time after changing the text. @thedoomshine glad you're tagged on this PR - maybe you have some advice!

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.4.3-alpha.ab8cde.0<br/>@codecademy/gamut-labs@18.3.3-alpha.ab8cde.0<br/>@codecademy/gamut@33.4.1-alpha.ab8cde.0<br/>@codecademy/styleguide@40.3.1-alpha.ab8cde.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://60b1428945107b00daeca047--gamut-preview.netlify.app>https://60b1428945107b00daeca047--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/60b1428945107b00daeca047>Deploy Logs</a></p>

File: packages/gamut/src/Form/TextArea.tsx [hunk 2]
```diff
@@ -28,24 +31,48 @@ const StyledTextArea = styled.textarea<TextAreaProps>`
...
+          activated={activated}
+          onChange={(event) => changeHandler(event)}
+        />
+        {characterMax && (
+          <FlexBox alignItems="flex-end" flexDirection="column">
+            <Text textColor={maxCharactersExceeded ? 'red' : 'navy'}>
+              ({characterCount} / {characterMax})
+            </Text>
+          </FlexBox>
+        )}
```
Comment (codecaaron) [inline packages/gamut/src/Form/TextArea.tsx:74]: I think we can avoid all custom code with just adding this as a validator https://react-hook-form.com/v6/api#register

```tsx
<GridForm 
  fields={[
    {
     name: "cool-textarea"
     validation: {
        validate: (input: string) => input.length > characterMax ? `${input.length} / ${charMax}` : undefined
      }
  }]
/>
```

It won't persist if we don't exceed the maximum - but we could optionally forward the error state to a persistent counter inside if it is specified :)

**Comment by katiezutter** [comment]
-- Putting this PR on hold for now, reaching back to design on some things! --

**Comment by katiezutter** [comment]
After chatting with Richie about this, we won't be implementing the character counter right now, and instead going with the solution that we currently have in production for the profiles work:
 
Screen Shot 2021-06-03 at 3 58 43 PM

@codecaaron @dreamwasp @jakemhiller - closing this one out!

**Pull request closed without merging**